libs/vkd3d/swapchain.c

```diff
@@ -2081,11 +2201,28 @@ static void dxgi_vk_swap_chain_present_iteration(struct dxgi_vk_swap_chain *chai
      * Non-FIFO swapchains will pump their frame latency handles through the fallback path of blit command being done.
      * Especially on Xwayland, the present ID is updated when images actually hit on-screen due to MAILBOX behavior.
      * This would unnecessarily stall our progress. */
-    if (chain->wait_thread.active && !chain->present.present_id_valid && swapchain_is_fifo)
+    if (chain->wait_thread.active && !chain->present.present_id_valid &&
+        (swapchain_is_fifo || chain->present.low_latency_state.mode))
     {
-        /* If we recreate swapchain, we still want to maintain a monotonically increasing counter here for
-         * profiling purposes. */
-        chain->present.present_id = chain->present.complete_count + 1;
+        if (chain->present.low_latency_state.mode &&
+            chain->request.low_latency_frame_id > chain->present.previous_application_frame_id)
+        {
+            chain->present.present_id = chain->request.low_latency_frame_id;
+            chain->present.previous_application_frame_id = chain->request.low_latency_frame_id;
+        }
+        else
+        {
+            /* If we recreate swapchain, we still want to maintain a monotonically increasing counter here for
+             * profiling purposes. */
+            chain->present.present_id = chain->present.low_latency_state.mode ?
+                chain->present.present_id + 1 : chain->present.complete_count + 1;
+
+            /* It is possible for an application to stop providing low latency frame ids. If that happens we are
+             * now responsible for ensuring the present id is always incrementing. If the application starts to
+             * provide them again, we will have to recreate the swapchain. */
+            chain->present.using_application_frame_id = false;
```

Comment:
Is this set to true somewhere if application frame ID *is* actually used?

Reply:
The state machine starts in "use app frame id state" when Reflex is enabled. When it observes a present without frame ID update, it switches to "don't use app frame id state". I suggested an alternative design in https://github.com/HansKristian-Work/vkd3d-proton/pull/1739#discussion_r1492468593.

Reply:
using_application_frame_id was previously set to true in dxgi_vk_swap_chain_recreate_swapchain_in_present_task if low latency mode was enabled. I have updated how the present id fallback works so this isn't relevant anymore though.

Reply:
The new code seems correct.